libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/MapboxRouteLineApi.kt

```diff
@@ -168,42 +173,66 @@ class MapboxRouteLineApi(
      * Updates which route is identified as the primary route.
      *
      * @param route the [DirectionsRoute] which should be designated as the primary
-     *
-     * @return a state which contains the side effects to be applied to the map displaying the
-     * newly designated route line.
+     * @param consumer a consumer of the result of calling this method which should be applied to
+     * a render method in [MapboxRouteLineView]
      */
-    fun updateToPrimaryRoute(route: DirectionsRoute): Expected<RouteSetValue, RouteLineError> {
-        val newRoutes = directionsRoutes.filter { it != route }.toMutableList().also {
-            it.add(0, route)
+    fun updateToPrimaryRoute(
+        route: DirectionsRoute,
+        consumer: MapboxNavigationConsumer<Expected<RouteSetValue, RouteLineError>>
+    ) {
+        jobControl.scope.launch {
+            mutex.withLock {
+                val newRoutes = directionsRoutes.filter { it != route }.toMutableList().also {
+                    it.add(0, route)
+                }
+                val featureDataProvider: () -> List<RouteFeatureData> =
+                    MapboxRouteLineUtils.getRouteFeatureDataProvider(newRoutes)
```

Comment:
Why do we re-decode the geometry from polyline to feature collection whenever an alternative is selected? The geometries don't change, we should be able to cache them. This looks to be one of the slowdowns in the alternative route line selection.

We should also execute it on a worker thread instead of the main thread, but that wouldn't resolve the unnecessary duplicate caluclations.

Reply:
We can treat this as a follow-up.

Reply:
For clarification MapboxRouteLineUtils.getRouteFeatureDataProvider(newRoutes) returns a function. The function is eventually invoked on a background thread in buildDrawRouteState().

```
        val routeFeatureDataDef = jobControl.scope.async(ThreadController.IODispatcher) {
            featureDataProvider()
        }
```